ethcore/sync/src/block_sync.rs

```diff
@@ -484,18 +483,19 @@ impl BlockDownloader {
 			let block = block_and_receipts.block;
 			let receipts = block_and_receipts.receipts;
 
-			// Perform basic block verification
-			if !Block::is_good(&block) {
-				debug!(target: "sync", "Bad block rlp: {:?}", block);
-				bad = true;
-				break;
-			}
-
-			let (h, number, parent) = {
-				let header = view!(BlockView, &block).header_view();
-				(header.hash(), header.number(), header.parent_hash())
+			let block = match Unverified::from_rlp(block) {
+				Ok(block) => block,
+				Err(_) => {
+					debug!(target: "sync", "Bad block rlp");
+					bad = true;
```

Comment:
This `bad`/ `if bad {…}` smells a little, especially in such a big function. How about just returning here and below with `Err(BlockDownloaderImportError::Invalid);`?

Reply:
@dvdplm we still need to update the number of imported blocks before returning.